File: crates/lunarbase-source-evm/src/fork.rs
//! Bounded fork resolution primitives for durable EVM event delivery.
//!
//! The quote plane deliberately does not own this window. Durable event
//! workers seed it from their block journal and invoke the HTTP resolver only
//! after a parent-link discontinuity.

use crate::rpc::backend::RpcHttpBackend;
use lunarbase_client::model::{BlockRef, Commitment, SourceError};
use lunarbase_math::B256;
use std::{collections::VecDeque, mem::size_of};
use thiserror::Error;

/// Count and retained-byte bounds for an unfinalized canonical window.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ForkWindowLimits {
    /// Maximum number of retained headers, including the finalized boundary.
    pub max_blocks: usize,
    /// Maximum bytes charged to retained headers.
    pub max_bytes: usize,
}

impl Default for ForkWindowLimits {
    fn default() -> Self {
        Self {
            max_blocks: 4096,
            max_bytes: 2 * 1024 * 1024,
        }
    }
}

/// Fail-closed fork-window or resolution error.
#[derive(Clone, Debug, Error, Eq, PartialEq)]
pub enum ForkError {
    /// A required block hash, parent hash, or coherent block identity is absent.
    #[error("incomplete or invalid block identity: {0}")]
    InvalidIdentity(String),
    /// A block does not extend the retained canonical tip.
    #[error("block is disconnected from the retained canonical window")]
    Disconnected,
    /// The configured count or byte budget is zero or too small.
    #[error("invalid canonical window limits")]
    InvalidLimits,
    /// The retained header count would exceed its hard limit.
    #[error("canonical window block budget exceeded")]
    BlockBudget,
    /// The retained header bytes would exceed their hard limit.
    #[error("canonical window byte budget exceeded")]
    ByteBudget,
    /// No common ancestor exists inside the retained window.
    #[error("common ancestor is outside the retained canonical window")]
    AncestorOutsideWindow,
    /// Resolving either branch would exceed the configured maximum depth.
    #[error("fork resolution depth exceeded")]
    DepthExceeded,
    /// A branch attempts to replace already-finalized history.
    #[error("fork conflicts with the finalized watermark")]
    FinalizedConflict,
    /// A resolution was computed against another canonical tip.
    #[error("fork resolution is stale")]
    StaleResolution,
    /// The exact block lookup failed.
    #[error(transparent)]
    Source(#[from] SourceError),
}

/// Deterministic correction plan computed without mutating canonical state.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ForkResolution {
    /// Last block shared by the abandoned and replacement branches.
    pub common_ancestor: BlockRef,
    /// Canonical tip against which this resolution was computed.
    pub old_tip: BlockRef,
    /// Proposed replacement tip.
    pub new_tip: BlockRef,
    /// Abandoned blocks in ascending height order, excluding the ancestor.
    pub old_branch: Vec<BlockRef>,
    /// Replacement blocks in ascending height order, excluding the ancestor.
    pub new_branch: Vec<BlockRef>,
}

impl ForkResolution {
    /// Returns true when both tips identify the same retained block.
    pub fn is_noop(&self) -> bool {
        self.old_branch.is_empty() && self.new_branch.is_empty()
    }
}

/// Amortized O(1) canonical header window without a per-block auxiliary index.
#[derive(Clone, Debug)]
pub struct CanonicalWindow {
    blocks: VecDeque<BlockRef>,
    finalized: Option<BlockRef>,
    limits: ForkWindowLimits,
    retained_bytes: usize,
}

impl CanonicalWindow {
    /// Creates an empty bounded window without preallocating its maximum size.
    pub fn new(limits: ForkWindowLimits) -> Result<Self, ForkError> {
        if limits.max_blocks == 0 || limits.max_bytes < block_charge() {
            return Err(ForkError::InvalidLimits);
        }
        Ok(Self {
            blocks: VecDeque::new(),
            finalized: None,
            limits,
            retained_bytes: 0,
        })
    }

    /// Number of retained blocks.
    pub fn len(&self) -> usize {
        self.blocks.len()
    }

    /// Returns whether the window is empty.
    pub fn is_empty(&self) -> bool {
        self.blocks.is_empty()
    }

    /// Conservative retained-byte charge used for admission control.
    pub fn retained_bytes(&self) -> usize {
        self.retained_bytes
    }

    /// Current canonical tip, if seeded.
    pub fn tip(&self) -> Option<&BlockRef> {
        self.blocks.back()
    }

    /// Highest finalized boundary retained by this window.
    pub fn finalized(&self) -> Option<&BlockRef> {
        self.finalized.as_ref()
    }

    /// Retained blocks from oldest to newest.
    pub fn blocks(&self) -> impl DoubleEndedIterator<Item = &BlockRef> {
        self.blocks.iter()
    }

    /// Appends one contiguous block. Exact duplicate tips are ignored.
    pub fn push_head(&mut self, block: BlockRef) -> Result<bool, ForkError> {
        validate_complete(&block)?;
        if let Some(tip) = self.tip() {
            validate_same_chain(tip, &block)?;
            if tip == &block {
                return Ok(false);
            }
            if block.cursor.block_number != tip.cursor.block_number.saturating_add(1)
                || block.parent_hash != tip.cursor.block_hash
            {
                return Err(ForkError::Disconnected);
            }
        }
        self.preflight(self.len().saturating_add(1))?;
        self.blocks.push_back(block);
        self.retained_bytes = self.retained_bytes.saturating_add(block_charge());
        Ok(true)
    }

    /// Replaces a same-height progressive tip sharing the same parent.
    pub fn replace_progressive_tip(&mut self, block: BlockRef) -> Result<(), ForkError> {
        validate_complete(&block)?;
        let Some(tip) = self.tip() else {
            return Err(ForkError::Disconnected);
        };
        validate_same_chain(tip, &block)?;
        if self
            .finalized
            .as_ref()
            .is_some_and(|finalized| finalized.cursor.block_number >= block.cursor.block_number)
        {
            return Err(ForkError::FinalizedConflict);
        }
        if tip.cursor.block_number != block.cursor.block_number
            || tip.parent_hash != block.parent_hash
        {
            return Err(ForkError::Disconnected);
        }
        *self.blocks.back_mut().expect("tip checked above") = block;
        Ok(())
    }

    /// Advances finality and prunes only history strictly older than the boundary.
    pub fn advance_finalized(&mut self, block: BlockRef) -> Result<(), ForkError> {
        validate_complete(&block)?;
        if block.cursor.commitment != Commitment::Finalized {
            return Err(ForkError::InvalidIdentity(
                "finalized watermark lacks finalized commitment".into(),
            ));
        }
        if let Some(previous) = &self.finalized {
            validate_same_chain(previous, &block)?;
            if block.cursor.block_number < previous.cursor.block_number
                || (block.cursor.block_number == previous.cursor.block_number
                    && block.cursor.block_hash != previous.cursor.block_hash)
            {
                return Err(ForkError::FinalizedConflict);
            }
        }
        let hash = required_hash(&block)?;
        let Some(index) = self.position(hash) else {
            return Err(ForkError::AncestorOutsideWindow);
        };
        let retained = &self.blocks[index];
        validate_same_chain(retained, &block)?;
        if retained.cursor.block_number != block.cursor.block_number
            || retained.cursor.execution_block_number != block.cursor.execution_block_number
            || retained.parent_hash != block.parent_hash
        {
            return Err(ForkError::InvalidIdentity(
                "finalized block does not match retained identity".into(),
            ));
        }
        self.blocks.drain(..index);
        self.retained_bytes = self.blocks.len().saturating_mul(block_charge());
        self.blocks[0] = block.clone();
        self.finalized = Some(block);
        Ok(())
    }

    /// Atomically switches the in-memory view after durable correction commits.
    pub fn apply_resolution(&mut self, resolution: &ForkResolution) -> Result<(), ForkError> {
        let Some(tip) = self.tip() else {
            return Err(ForkError::StaleResolution);
        };
        if tip != &resolution.old_tip {
            return Err(ForkError::StaleResolution);
        }
        let ancestor_hash = required_hash(&resolution.common_ancestor)?;
        let Some(ancestor_index) = self.position(ancestor_hash) else {
            return Err(ForkError::StaleResolution);
        };
        if self.blocks[ancestor_index] != resolution.common_ancestor
            || !self
                .blocks
                .iter()
                .skip(ancestor_index + 1)
                .eq(&resolution.old_branch)
        {
            return Err(ForkError::StaleResolution);
        }
        self.validate_replacement(resolution)?;
        let next_len = ancestor_index
            .saturating_add(1)
            .saturating_add(resolution.new_branch.len());
        self.preflight(next_len)?;

        self.blocks.truncate(ancestor_index + 1);
        self.blocks.extend(resolution.new_branch.iter().cloned());
        self.retained_bytes = self.blocks.len().saturating_mul(block_charge());
        Ok(())
    }

    fn validate_replacement(&self, resolution: &ForkResolution) -> Result<(), ForkError> {
        if self.finalized.as_ref().is_some_and(|finalized| {
            resolution.common_ancestor.cursor.block_number < finalized.cursor.block_number
        }) {
            return Err(ForkError::FinalizedConflict);
        }
        let mut parent = &resolution.common_ancestor;
        for block in &resolution.new_branch {
            validate_complete(block)?;
            validate_same_chain(parent, block)?;
            if block.cursor.block_number != parent.cursor.block_number.saturating_add(1)
                || block.parent_hash != parent.cursor.block_hash
            {
                return Err(ForkError::Disconnected);
            }
            parent = block;
        }
        if parent != &resolution.new_tip {
            return Err(ForkError::StaleResolution);
        }
        Ok(())
    }

    fn preflight(&self, blocks: usize) -> Result<(), ForkError> {
        if blocks > self.limits.max_blocks {
            return Err(ForkError::BlockBudget);
        }
        if blocks.saturating_mul(block_charge()) > self.limits.max_bytes {
            return Err(ForkError::ByteBudget);
        }
        Ok(())
    }

    fn position(&self, hash: B256) -> Option<usize> {
        self.blocks
            .iter()
            .position(|block| block.cursor.block_hash == Some(hash))
    }
}

/// Bounded exact-hash walker used only after a head-link discontinuity.
#[derive(Clone)]
pub struct ForkResolver {
    backend: RpcHttpBackend,
    max_depth: usize,
}

impl ForkResolver {
    /// Creates a resolver with a strict maximum depth on both branches.
    pub fn new(backend: RpcHttpBackend, max_depth: usize) -> Result<Self, ForkError> {
        if max_depth == 0 {
            return Err(ForkError::InvalidLimits);
        }
        Ok(Self { backend, max_depth })
    }

    /// Resolves the backend's configured canonical tip with parent linkage.
    pub async fn canonical_tip(&self) -> Result<BlockRef, ForkError> {
        self.backend
            .snapshot_block_ref(self.backend.network())
            .await
            .map_err(Into::into)
    }

    /// Resolves one exact block hash with network-specific execution context.
    pub async fn block_ref_by_hash(
        &self,
        block_hash: B256,
        commitment: Commitment,
    ) -> Result<BlockRef, ForkError> {
        self.backend
            .block_ref_by_hash(block_hash, commitment)
            .await
            .map_err(Into::into)
    }

    /// Resolves the provider's finalized watermark with parent linkage.
    pub async fn finalized_tip(&self) -> Result<BlockRef, ForkError> {
        self.backend
            .block_ref_at_tag("finalized", Commitment::Finalized)
            .await
            .map_err(Into::into)
    }

    /// Computes a correction plan without mutating the retained window.
    pub async fn resolve(
        &self,
        window: &CanonicalWindow,
        new_tip: BlockRef,
    ) -> Result<ForkResolution, ForkError> {
        validate_complete(&new_tip)?;
        let Some(old_tip) = window.tip().cloned() else {
            return Err(ForkError::AncestorOutsideWindow);
        };
        validate_same_chain(&old_tip, &new_tip)?;
        if new_tip.cursor.chain_id != self.backend.chain_id() {
            return Err(ForkError::InvalidIdentity(
                "resolver chain id mismatch".into(),
            ));
        }
        if old_tip == new_tip {
            return Ok(ForkResolution {
                common_ancestor: old_tip.clone(),
                old_tip: old_tip.clone(),
                new_tip: old_tip,
                old_branch: Vec::new(),
                new_branch: Vec::new(),
            });
        }
        if window.finalized().is_some_and(|finalized| {
            new_tip.cursor.block_number <= finalized.cursor.block_number
                && new_tip.cursor.block_hash != finalized.cursor.block_hash
        }) {
            return Err(ForkError::FinalizedConflict);
        }

        let mut descending = vec![new_tip.clone()];
        let (common_ancestor, ancestor_index) = loop {
            let child = descending.last().expect("new tip seeded above");
            let parent_hash = required_parent_hash(child)?;
            if let Some(index) = window.position(parent_hash) {
                let ancestor = window.blocks[index].clone();
                if child.cursor.block_number != ancestor.cursor.block_number.saturating_add(1) {
                    return Err(ForkError::Disconnected);
                }
                break (ancestor, index);
            }
            if descending.len() >= self.max_depth {
                return Err(ForkError::DepthExceeded);
            }
            if window.finalized().is_some_and(|finalized| {
                child.cursor.block_number <= finalized.cursor.block_number.saturating_add(1)
            }) {
                return Err(ForkError::FinalizedConflict);
            }
            if child.cursor.block_number == 0 {
                return Err(ForkError::AncestorOutsideWindow);
            }
            let parent = self
                .backend
                .block_ref_by_hash(parent_hash, child.cursor.commitment)
                .await?;
            validate_complete(&parent)?;
            if parent.cursor.block_number.saturating_add(1) != child.cursor.block_number {
                return Err(ForkError::Disconnected);
            }
            descending.push(parent);
        };

        let old_branch = window
            .blocks
            .iter()
            .skip(ancestor_index + 1)
            .cloned()
            .collect::<Vec<_>>();
        if old_branch.len() > self.max_depth {
            return Err(ForkError::DepthExceeded);
        }
        descending.reverse();
        Ok(ForkResolution {
            common_ancestor,
            old_tip,
            new_tip,
            old_branch,
            new_branch: descending,
        })
    }
}

const fn block_charge() -> usize {
    size_of::<BlockRef>()
}

fn required_hash(block: &BlockRef) -> Result<B256, ForkError> {
    block
        .cursor
        .block_hash
        .ok_or_else(|| ForkError::InvalidIdentity("block hash is absent".into()))
}

fn required_parent_hash(block: &BlockRef) -> Result<B256, ForkError> {
    block
        .parent_hash
        .ok_or_else(|| ForkError::InvalidIdentity("parent hash is absent".into()))
}

fn validate_complete(block: &BlockRef) -> Result<(), ForkError> {
    required_hash(block)?;
    required_parent_hash(block)?;
    if block.cursor.transaction_index.is_some() || block.cursor.log_index.is_some() {
        return Err(ForkError::InvalidIdentity(
            "block reference contains event coordinates".into(),
        ));
    }
    Ok(())
}

fn validate_same_chain(left: &BlockRef, right: &BlockRef) -> Result<(), ForkError> {
    if left.cursor.chain_id != right.cursor.chain_id {
        return Err(ForkError::InvalidIdentity("chain id mismatch".into()));
    }
    Ok(())
}

File: packages/source-evm/src/fork.ts
/** Bounded fork resolution primitives for durable EVM event delivery. */
import { Commitment, type BlockRef, type ChainCursor } from "@lunarbase-lab/pmm-v2-client";
import type * as Hex from "ox/Hex";
import { RpcError, RpcHttpBackend, parseHash } from "./rpc.js";

/** Conservative charge including two objects, bigints, and both hash strings. */
export const BLOCK_REF_RETAINED_BYTES = 512;

/** Count and retained-byte bounds for an unfinalized canonical window. */
export interface ForkWindowLimits {
  readonly maxBlocks: number;
  readonly maxBytes: number;
}

/** Default fork window bounds. */
export const DEFAULT_FORK_WINDOW_LIMITS: ForkWindowLimits = Object.freeze({
  maxBlocks: 4096,
  maxBytes: 2 * 1024 * 1024,
});

/** Stable fail-closed fork error categories. */
export type ForkErrorCode =
  | "INVALID_IDENTITY"
  | "DISCONNECTED"
  | "INVALID_LIMITS"
  | "BLOCK_BUDGET"
  | "BYTE_BUDGET"
  | "ANCESTOR_OUTSIDE_WINDOW"
  | "DEPTH_EXCEEDED"
  | "FINALIZED_CONFLICT"
  | "STALE_RESOLUTION"
  | "RPC";

/** Fail-closed fork-window or resolution error. */
export class ForkError extends Error {
  constructor(
    readonly code: ForkErrorCode,
    message: string,
    options?: ErrorOptions,
  ) {
    super(message, options);
    this.name = "ForkError";
  }
}

/** Deterministic correction plan computed without mutating canonical state. */
export interface ForkResolution {
  readonly commonAncestor: BlockRef;
  readonly oldTip: BlockRef;
  readonly newTip: BlockRef;
  /** Abandoned blocks in ascending height order, excluding the ancestor. */
  readonly oldBranch: readonly BlockRef[];
  /** Replacement blocks in ascending height order, excluding the ancestor. */
  readonly newBranch: readonly BlockRef[];
}

/**
 * Bounded canonical header window.
 *
 * Normal append is O(1). Finality advances a start index instead of using
 * `Array.shift()` and compacts only occasionally outside the ingestion path.
 */
export class CanonicalWindow {
  private storage: BlockRef[] = [];
  private start = 0;
  private finalizedBlock?: BlockRef;
  readonly limits: ForkWindowLimits;

  constructor(limits: ForkWindowLimits = DEFAULT_FORK_WINDOW_LIMITS) {
    if (
      !positiveInteger(limits.maxBlocks) ||
      !positiveInteger(limits.maxBytes) ||
      limits.maxBytes < BLOCK_REF_RETAINED_BYTES
    )
      throw new ForkError("INVALID_LIMITS", "canonical window limits must be positive");
    this.limits = Object.freeze({ ...limits });
  }

  /** Number of retained blocks. */
  get length(): number {
    return this.storage.length - this.start;
  }

  /** Conservative retained-byte charge used for admission control. */
  get retainedBytes(): number {
    return this.length * BLOCK_REF_RETAINED_BYTES;
  }

  /** Current canonical tip, if seeded. */
  get tip(): BlockRef | undefined {
    return this.storage.at(-1);
  }

  /** Highest finalized boundary retained by this window. */
  get finalized(): BlockRef | undefined {
    return this.finalizedBlock;
  }

  /** Iterates retained blocks from oldest to newest without copying them. */
  *blocks(): IterableIterator<BlockRef> {
    for (let index = this.start; index < this.storage.length; index += 1) yield this.storage[index]!;
  }

  /** Appends one contiguous block. Exact duplicate tips are ignored. */
  pushHead(input: BlockRef): boolean {
    const block = ownCompleteBlock(input);
    const tip = this.tip;
    if (tip !== undefined) {
      validateSameChain(tip, block);
      if (sameBlock(tip, block)) return false;
      if (block.cursor.blockNumber !== tip.cursor.blockNumber + 1n || block.parentHash !== tip.cursor.blockHash)
        throw new ForkError("DISCONNECTED", "block is disconnected from the retained canonical window");
    }
    this.preflight(this.length + 1);
    this.storage.push(block);
    return true;
  }

  /** Replaces a same-height progressive tip sharing the same parent. */
  replaceProgressiveTip(input: BlockRef): void {
    const block = ownCompleteBlock(input);
    const tip = this.tip;
    if (tip === undefined) throw new ForkError("DISCONNECTED", "canonical window is empty");
    validateSameChain(tip, block);
    if (this.finalizedBlock !== undefined && this.finalizedBlock.cursor.blockNumber >= block.cursor.blockNumber)
      throw new ForkError("FINALIZED_CONFLICT", "progressive head would replace finalized history");
    if (tip.cursor.blockNumber !== block.cursor.blockNumber || tip.parentHash !== block.parentHash)
      throw new ForkError("DISCONNECTED", "progressive head does not share the retained tip parent");
    this.storage[this.storage.length - 1] = block;
  }

  /** Advances finality and prunes only history strictly older than the boundary. */
  advanceFinalized(input: BlockRef): void {
    const block = ownCompleteBlock(input);
    if (block.cursor.commitment !== Commitment.Finalized)
      throw new ForkError("INVALID_IDENTITY", "finalized watermark lacks finalized commitment");
    const previous = this.finalizedBlock;
    if (previous !== undefined) {
      validateSameChain(previous, block);
      if (
        block.cursor.blockNumber < previous.cursor.blockNumber ||
        (block.cursor.blockNumber === previous.cursor.blockNumber &&
          block.cursor.blockHash !== previous.cursor.blockHash)
      )
        throw new ForkError("FINALIZED_CONFLICT", "finalized watermark regressed or changed hash");
    }
    const index = this.position(requiredHash(block));
    if (index < 0) throw new ForkError("ANCESTOR_OUTSIDE_WINDOW", "finalized block is outside the retained window");
    const retained = this.storage[index]!;
    validateSameChain(retained, block);
    if (
      retained.cursor.blockNumber !== block.cursor.blockNumber ||
      retained.cursor.executionBlockNumber !== block.cursor.executionBlockNumber ||
      retained.parentHash !== block.parentHash
    )
      throw new ForkError("INVALID_IDENTITY", "finalized block does not match retained identity");
    this.start = index;
    this.storage[index] = block;
    this.finalizedBlock = block;
    this.compactAfterFinality();
  }

  /** Atomically switches the view after its durable correction commits. */
  applyResolution(resolution: ForkResolution): void {
    const tip = this.tip;
    if (tip === undefined || !sameBlock(tip, resolution.oldTip))
      throw new ForkError("STALE_RESOLUTION", "fork resolution was computed against another tip");
    const ancestorIndex = this.position(requiredHash(resolution.commonAncestor));
    if (ancestorIndex < 0 || !sameBlock(this.storage[ancestorIndex]!, resolution.commonAncestor))
      throw new ForkError("STALE_RESOLUTION", "common ancestor is no longer retained");
    const currentOldBranch = this.storage.slice(ancestorIndex + 1);
    if (!sameBranch(currentOldBranch, resolution.oldBranch))
      throw new ForkError("STALE_RESOLUTION", "abandoned branch no longer matches the window");
    const replacement = resolution.newBranch.map(ownCompleteBlock);
    this.validateReplacement(resolution.commonAncestor, resolution.newTip, replacement);
    this.preflight(ancestorIndex - this.start + 1 + replacement.length);
    this.storage.length = ancestorIndex + 1;
    this.storage.push(...replacement);
  }

  private validateReplacement(commonAncestor: BlockRef, newTip: BlockRef, branch: readonly BlockRef[]): void {
    if (this.finalizedBlock !== undefined && commonAncestor.cursor.blockNumber < this.finalizedBlock.cursor.blockNumber)
      throw new ForkError("FINALIZED_CONFLICT", "fork crosses the finalized watermark");
    let parent = commonAncestor;
    for (const block of branch) {
      validateSameChain(parent, block);
      if (block.cursor.blockNumber !== parent.cursor.blockNumber + 1n || block.parentHash !== parent.cursor.blockHash)
        throw new ForkError("DISCONNECTED", "replacement branch contains a parent discontinuity");
      parent = block;
    }
    if (!sameBlock(parent, newTip))
      throw new ForkError("STALE_RESOLUTION", "replacement branch does not end at new tip");
  }

  private preflight(blocks: number): void {
    if (blocks > this.limits.maxBlocks) throw new ForkError("BLOCK_BUDGET", "canonical window block budget exceeded");
    if (blocks * BLOCK_REF_RETAINED_BYTES > this.limits.maxBytes)
      throw new ForkError("BYTE_BUDGET", "canonical window byte budget exceeded");
  }

  private position(hash: Hex.Hex): number {
    for (let index = this.start; index < this.storage.length; index += 1)
      if (this.storage[index]!.cursor.blockHash === hash) return index;
    return -1;
  }

  private compactAfterFinality(): void {
    if (this.start < 1024 || this.start * 2 < this.storage.length) return;
    this.storage = this.storage.slice(this.start);
    this.start = 0;
  }

  /** Internal exact-hash lookup used by the rare fork resolver. */
  findByHash(hash: Hex.Hex): { readonly block: BlockRef; readonly index: number } | undefined {
    const index = this.position(hash);
    return index < 0 ? undefined : { block: this.storage[index]!, index };
  }

  /** Internal branch copy used only while constructing a correction. */
  branchAfter(index: number): readonly BlockRef[] {
    return this.storage.slice(index + 1);
  }
}

/** Bounded exact-hash walker used only after a head-link discontinuity. */
export class ForkResolver {
  constructor(
    private readonly backend: RpcHttpBackend,
    private readonly maxDepth: number,
  ) {
    if (!positiveInteger(maxDepth)) throw new ForkError("INVALID_LIMITS", "fork depth must be positive");
  }

  /** Computes a correction plan without mutating the retained window. */
  async resolve(window: CanonicalWindow, input: BlockRef): Promise<ForkResolution> {
    const newTip = ownCompleteBlock(input);
    const oldTip = window.tip;
    if (oldTip === undefined) throw new ForkError("ANCESTOR_OUTSIDE_WINDOW", "canonical window is empty");
    validateSameChain(oldTip, newTip);
    if (newTip.cursor.chainId !== this.backend.chainId)
      throw new ForkError("INVALID_IDENTITY", "resolver chain id mismatch");
    if (sameBlock(oldTip, newTip))
      return { commonAncestor: oldTip, oldTip, newTip: oldTip, oldBranch: [], newBranch: [] };
    const finalized = window.finalized;
    if (
      finalized !== undefined &&
      newTip.cursor.blockNumber <= finalized.cursor.blockNumber &&
      newTip.cursor.blockHash !== finalized.cursor.blockHash
    )
      throw new ForkError("FINALIZED_CONFLICT", "replacement tip conflicts with finalized history");

    const descending: BlockRef[] = [newTip];
    let commonAncestor: BlockRef;
    let ancestorIndex: number;
    for (;;) {
      const child = descending[descending.length - 1]!;
      const parentHash = requiredParentHash(child);
      const retained = window.findByHash(parentHash);
      if (retained !== undefined) {
        if (child.cursor.blockNumber !== retained.block.cursor.blockNumber + 1n)
          throw new ForkError("DISCONNECTED", "replacement parent height is inconsistent");
        commonAncestor = retained.block;
        ancestorIndex = retained.index;
        break;
      }
      if (descending.length >= this.maxDepth) throw new ForkError("DEPTH_EXCEEDED", "fork resolution depth exceeded");
      if (finalized !== undefined && child.cursor.blockNumber <= finalized.cursor.blockNumber + 1n)
        throw new ForkError("FINALIZED_CONFLICT", "fork crosses the finalized watermark");
      if (child.cursor.blockNumber === 0n)
        throw new ForkError("ANCESTOR_OUTSIDE_WINDOW", "common ancestor is outside the retained window");
      let parent: BlockRef;
      try {
        parent = ownCompleteBlock(await this.backend.blockRefByHash(parentHash, child.cursor.commitment));
      } catch (error) {
        if (error instanceof ForkError) throw error;
        const detail = error instanceof Error ? error.message : String(error);
        throw new ForkError("RPC", `exact-hash fork lookup failed: ${detail}`, { cause: error });
      }
      if (parent.cursor.blockNumber + 1n !== child.cursor.blockNumber)
        throw new ForkError("DISCONNECTED", "replacement branch height is inconsistent");
      descending.push(parent);
    }

    const oldBranch = window.branchAfter(ancestorIndex);
    if (oldBranch.length > this.maxDepth) throw new ForkError("DEPTH_EXCEEDED", "abandoned branch is too deep");
    return {
      commonAncestor,
      oldTip,
      newTip,
      oldBranch,
      newBranch: descending.reverse(),
    };
  }
}

function positiveInteger(value: number): boolean {
  return Number.isSafeInteger(value) && value > 0;
}

function ownCompleteBlock(block: BlockRef): BlockRef {
  const blockHash = requiredHash(block);
  const parentHash = requiredParentHash(block);
  if (block.cursor.transactionIndex !== undefined || block.cursor.logIndex !== undefined)
    throw new ForkError("INVALID_IDENTITY", "block reference contains event coordinates");
  const cursor: ChainCursor = Object.freeze({ ...block.cursor, blockHash });
  return Object.freeze({ cursor, parentHash });
}

function requiredHash(block: BlockRef): Hex.Hex {
  if (block.cursor.blockHash === undefined) throw new ForkError("INVALID_IDENTITY", "block hash is absent");
  try {
    return parseHash(block.cursor.blockHash, "block hash");
  } catch (error) {
    throw new ForkError("INVALID_IDENTITY", "block hash is invalid", { cause: error });
  }
}

function requiredParentHash(block: BlockRef): Hex.Hex {
  if (block.parentHash === undefined) throw new ForkError("INVALID_IDENTITY", "parent hash is absent");
  try {
    return parseHash(block.parentHash, "parent hash");
  } catch (error) {
    throw new ForkError("INVALID_IDENTITY", "parent hash is invalid", { cause: error });
  }
}

function validateSameChain(left: BlockRef, right: BlockRef): void {
  if (left.cursor.chainId !== right.cursor.chainId) throw new ForkError("INVALID_IDENTITY", "block chain id mismatch");
}

function sameBlock(left: BlockRef, right: BlockRef): boolean {
  return (
    left.cursor.chainId === right.cursor.chainId &&
    left.cursor.blockNumber === right.cursor.blockNumber &&
    left.cursor.executionBlockNumber === right.cursor.executionBlockNumber &&
    left.cursor.blockHash === right.cursor.blockHash &&
    left.cursor.commitment === right.cursor.commitment &&
    left.cursor.sourceSequence === right.cursor.sourceSequence &&
    left.cursor.sourceSubIndex === right.cursor.sourceSubIndex &&
    left.parentHash === right.parentHash
  );
}

function sameBranch(left: readonly BlockRef[], right: readonly BlockRef[]): boolean {
  return left.length === right.length && left.every((block, index) => sameBlock(block, right[index]!));
}

/** Returns true when a correction leaves the retained tip unchanged. */
export function isNoopResolution(resolution: ForkResolution): boolean {
  return resolution.oldBranch.length === 0 && resolution.newBranch.length === 0;
}

/** Narrows transport failures when callers need separate health metrics. */
export function isForkRpcError(error: unknown): boolean {
  return error instanceof ForkError && error.code === "RPC" && error.cause instanceof RpcError;
}
